x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/InternalUsers.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.user;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class InternalUsers {
+
+    private record UserInstance(String name, User user, @Nullable RoleDescriptor role, Predicate<User> isUserPredicate) {
+        private UserInstance {
+            // Check that the parameters align
+            assert User.isInternal(user) : "User " + user + " is not internal";
```

Comment:
There would be a circular reference, but I could fix that.

My preference it to get rid of `User.isInternal`, but it's called from a lot of places so it would be better as a standalone PR.

Reply:
> My preference it to get rid of User.isInternal, but it's called from a lot of places so it would be better as a standalone PR.

That's my preference as well. Also agree it should be a separate PR.

If it's trickly to fix the circular reference, don't bother with it given we plan to get rid of `User.isInternal`. Maybe leave a TODO somewhere so we don't forget.